boot-speed/clouds/measure-cloud.py

```diff
@@ -0,0 +1,242 @@
+#!/usr/bin/env python3
+"""
+Measure the boot speed of cloud instances.
+
+Copyright 2019 Canonical Ltd.
+Paride Legovini <paride.legovini@canonical.com>
+"""
+
+import argparse
+import datetime as dt
+import glob
+import json
+import logging
+import tarfile
+import os
+import paramiko
+import pycloudlib
+import sys
+
+from pathlib import Path
+
+
+known_clouds = ['ec2', 'gce']
+
+
+def create_sftp_client(host, user, port=22, password=None, keyfilepath=None):
+    """
+    create_sftp_client(host, port, user, password, keyfilepath) -> SFTPClient
+    """
+    sftp = None
+    key = None
+    transport = None
+    try:
+        if keyfilepath is not None:
+            # Get private key used to authenticate user.
+            key = paramiko.RSAKey.from_private_key_file(keyfilepath)
+
+        # Create Transport object using supplied method of authentication.
+        transport = paramiko.Transport((host, port))
+        transport.connect(None, user, password, key)
+
+        sftp = paramiko.SFTPClient.from_transport(transport)
+
+        return sftp
+    except Exception as e:
+        print('An creating SFTP client: %s: %s' % (e.__class__, e))
+        if sftp is not None:
+            sftp.close()
+        if transport is not None:
+            transport.close()
+        pass
+
+
+def cloudid2serial(id):
+    stream = pycloudlib.streams.Streams(
+        mirror_url='https://cloud-images.ubuntu.com/daily',
+        keyring_path='/usr/share/keyrings/ubuntu-cloudimage-keyring.gpg'
+    )
+    filters = ['id=%s' % id]
+    res = stream.query(filters)
+    id = res[0]['id']
+    return id
+
+
+def measure_instance(instance, datadir, reboots=1):
```

Comment:
the first part of this function is reused code from devices.

Can we create one function that has different calls depending on the type?

Although, only devices will auto-reboot, so cloud and lxd/kvm should not, so maybe we can remove all of that code as well to make this simpler. Thoughts?

Reply:
I can see your point, but I'm not sure it's worth it. The two cases (testflinger devices vs. clouds) are indeed similar but different enough to require to be handled differently, at least in part. The auto-reboot thing is one example. The reused commands are just a few, while the common part is in `bootspeed.sh`, which is exactly the same for devices, clouds, or any other system we may want to measure.

A couple of commands could be moved there, e.g. the boot_id check. The `rm -rf artifacts` is redundant (bootspeed.sh already does it), and it's there just for extra safety, as we want to be really sure not to retrieve and old measurement. What remains is very little.